## MBA em Ciência de Dados
# Redes Neurais e Arquiteturas Profundas

### <span style="color:darkred">Módulo I - Deep Learning e redes do tipo Perceptron</span>


### <span style="color:darkred">Exercícios</span>

Moacir Antonelli Ponti

USP

---

#### <span style="color:red">Recomenda-se fortemente que os exercícios sejam feitos sem consultar as respostas antecipadamente.</span>

---

# Parte 1 - Exercícios Essenciais

---

### Exercício 1)

As redes neurais artificiais passaram por um chamado "inverno" em que recebiam pouca atenção da academia e do mercado quando comparado aos sistemas especialistas. Porque isso mudou em meados de 2010?

 (a) Devido a descobertas da inteligência artificial em 2012 que revolucionou a área<br>
 <font color='red'>(b) Trabalhos científicos mostraram que a utilização de grandes bases de dados anotadas e o uso de placas gráficas permitiam treinar modelos com baixo erro de classificação</font><br>
 (c) Devido a melhoria dos sistemas computacionais, incluindo melhores processadores (CPUs), maior memória principal disponível (RAM) e placas gráficas (GPUs)<br>
 (d) Pois foi nessa época em que a área de ciência de dados tomou destaque, devido a grande quantidade de dados disponível (big data)<br>



O uso de processamento gráfico paralelo (GPU) permitiu o uso de grandes quantidades de dados para treinamento e a consequentemente a obtenção de melhores resultados com modelos de redes neurais.

---
### Exercício 2)

Seja $\mathbf{x} \in X$ um vetor com informações sobre perfis de clientes de um e-commerce, incluindo: data de nascimento, cidade, quantidade de pedidos realizados nos últimos 12 meses, total em reais das compras, entre outros. Desejamos obter um modelo que seja capaz de receber por entrada uma instância de $X$ e inferir a probabilidade $y$ de um cliente realizar uma compra no próximo mês. Considere duas formulações conforme abaixo:

1. Modelo A: $f(\mathbf{x}) = \hat{y}$
2. Modelo B: $$\begin{align}
g(\mathbf{x}) &= \mathbf{r},\\
h(\mathbf{r}) &= \mathbf{s},\\
i(\mathbf{s}) &= \mathbf{w},\\
f(\mathbf{w}) &= \hat{y}
\end{align}
$$

O que podemos dizer sobre A e B?

(a) A é um classificador linear e B é um regressor multi-estágio<br>
(b) apenas B é uma formulação de aprendizado de máquina, enquanto A não pode ser considerado aprendizado de máquina<br>
(c) A é um classificador linear como o Perceptron, enquanto B é uma rede neural profunda<br>
<font color='red'>(d) A é uma formulação rasa, enquanto que B é uma formulação profunda para aprendizado de máquina</font><br>

---

### Exercício 3)

Seja $\mathbf{x}$ um vetor de entrada e $s$ um valor de saída de um neurônio de uma rede neural baseada em Perceptron. O processamento pode ser escrito na forma:

$f(\mathbf{x}) = a(\mathbf{w}^t\mathbf{x}+b) = s$,

sendo que $a()$ é a função de ativação. Sabendo que a entrada tem $d$ dimensões, quais são as dimensões das variáveis $w$ e $b$?

<font color='red'>(a) $w$ é um vetor com $d$ dimensões, enquanto $b$ é escalar possuindo uma dimensão</font><br>
(b) $w$ é um vetor com 2 dimensões, enquanto $b$ é escalar possuindo uma dimensão<br>
(c) $w$ é escalar com 1 dimensão, enquanto $b$ é o vetor bias cujo número de dimensões é igual ao número de classes do problema<br>
(d) $w$ é um vetor com $d-1$ dimensões, enquanto $b$ é bias e portanto possui apenas 1 dimensão


w é o vetor de pesos que multiplicarão cada atributo, portanto sua dimensão deve ser igual à dos atributos, portanto e dimensões<br>
b é um escalar(bias)<br>

---

### Exercício 4)

Seja $\mathbf{x}$ um vetor de entrada e $\mathbf{s}$ um vetor de saída de uma camada de rede neural baseada em Perceptron. O processamento pode ser escrito na forma:

$f(\mathbf{x}) = a(W\mathbf{x}+\mathbf{b}) = \mathbf{s}$,

sendo que $a()$ é a função de ativação. Sabendo que a entrada tem $d$ dimensões e a saída tem $z$ dimensões, quais são as dimensões das variáveis $W$ e $b$ e quantos neurônios possui essa camada?

(a) são $d$ neurônios, sendo que $W$ possui $z \times d$, e $b$ possui $d$ dimensões<br>
(b) são $d \times z$ neurônios, sendo que $W$ possui $z \times d$, e $b$ possui $d$ dimensões<br>
<font color='red'>(c) são $z$ neurônios, sendo que $W$ possui $z \times d$, e $b$ possui $z$ dimensões</font><br>
(d) são $z$ neurônios, sendo que $W$ possui $d$, e $b$ possui $z$ dimensões


É necessário 1 neurônio para cada dimensão de saída, portanto, são z neurônios.<br>
W é o vetor de pesos, portanto d dimensões multiplicada pelo número de dimensões de saída.<br>
b (bias) é um escalar possuindo 1 dimensão para cada dimensão de saída, portanto, z dimensões.<br>

---

### Exercício 5)

Qual a diferença entre as funções de ativação *sigmoide* e *softmax* aplicadas a uma camada de uma rede neural?

(a) a sigmóide produz como saída vetores cuja somatória é 1 e que podem ser interpretados como uma distribuição de probabilidade, enquanto a softmax comprime os valores da saída entre 0 e 1 <br>
(b) a sigmóide produz como saída vetores cujos valores estão entre -1 e 1, enquanto a softmax comprime os valores da saída entre 0 e 1<br>
<font color='red'>(c) a sigmóide comprime os valores da saída entre 0 e 1, enquanto a softmax produz como saída vetores cuja somatória é 1 e que podem ser interpretados como uma distribuição de probabilidade</font><br>
(d) a sigmóide produz como saída vetores cujos valores individuais estão entre 0 e 1 utilizando uma função suave, enquanto a softmax produz vetores no formato one-hot-encoding compatíveis


Sigmóide é uma função de ativação que transforma um sinal de entrada em um sinal de saída entre 0 e 1.<br>
Softmax é classificador que normaliza a saída de forma a somar 1, permitindo interpretar cada valor como sendo uma probabilidade e a saída pode ser descrita como uma função de distribuição de probabilidades de uma determinada classe.<br>

---

### Exercício 6)

Utilizando a biblioteca Keras, formule um modelo de rede neural sequencial, do tipo MLP, cuja entrada seja preparada para receber um vetor de 100 dimensões e que possua 2 camadas ocultas, cada uma com 50 neurônios, e 1 camada de saída, com 3 neurônios e função de ativação softmax.

Quantos parâmetros, no total, essa rede possui?

(a) 203<br>
(b) 10303<br>
(c) 7651<br>
<font color='red'>(d) 7753</font><br>

In [18]:
### seu codigo
from tensorflow import keras

model1 = keras.Sequential(
    [
        keras.layers.Dense(50, activation="relu", input_shape=(100,)),
        keras.layers.Dense(50, activation="relu"),
        keras.layers.Dense(3, activation="softmax"),
    ]
)
model1.summary()


Model: "sequential_4"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_16 (Dense)             (None, 50)                5050      
_________________________________________________________________
dense_17 (Dense)             (None, 50)                2550      
_________________________________________________________________
dense_18 (Dense)             (None, 3)                 153       
Total params: 7,753
Trainable params: 7,753
Non-trainable params: 0
_________________________________________________________________


O total de parâmetros a serem estimados é a soma de cada camada: 5050+2550+153 = 7753

---

### Exercício 7)

Considerando que temos uma base de dados de treinamento com 8000 exemplos. Definindo o tamanho do lote (minibatch size) como sendo 50, quantas iterações o algoritmo backpropagation deverá executar, adaptando os parâmetros da rede, para completar 5 épocas?

<font color='red'>(a) 800</font><br>
(a) 160<br>
(b) 801<br>
(c) 250<br>

In [8]:
N = 8000
batch_size = 50

iteracoes_epoca = N/batch_size
iteracoes_epoca * 5

800.0

Precisará de 8000/50 = 160 iterações para completar 1 época.<br>
Para completar 5 épocas = 160 x 5 = 800

# Parte 2 - Exercícios Complementares

### Exercício 8)

As redes neurais artificiais são uma família de métodos conhecidos por conexionistas. Assinale a alternativa mais completa que explica o porquê.
 
 <font color='red'>(a) Porque são inspirados por neurônios biológicos, que representam a informação por uma rede de unidades de processamento e o aprendizado pela força de suas conexões</font><br>
 (b) Porque apenas as redes neurais possuem uma estrutura em que há nós conectados pelos quais passam as informações<br>
 (c) Porque é possível conectar os neurônios artificiais de qualquer maneira, simulando um cérebro humano<br>
 (d) Porque são métodos de aprendizado de máquina supervisionados não-paramétricos, muito utilizados em tarefas de classificação e regressão, e que armazenam informações em nós organizados de forma hierárquica, desde sua entrada, conhecida por raiz, até a saída, chamada de folha.<br>

É chamado de método conexionista porque estabelece a força das conexões (vetor w)

---

### Exercício 9)

Defina as sementes aleatórias do numpy para 1 e do tensorflow para 2, depois carregue a base de dados boston housing da biblioteca Keras, conforme código abaixo. 

O objetivo dessa base de dados é obter a regressão do preço das casas com base em 13 características de entrada. Assim, os valores alvo (target) são escalares, tipicamente entre 10 e 50 (representando os preços em milhares de dólares).

Considerando o valor alvo descrito, e que você tem disponíveis as funções de ativação: sigmóide, relu, tangente hiperbólica e softmax, escolha as funções mais adequadas e monte uma arquitetura da rede com 3 camadas ocultas, todas com 32 neurônios.

Como deve ficar a arquitetura da rede neural?

(a) 4 camadas com 32, 32, 32 e 50 neurônios, 4202 parâmetros e função de ativação relu na camada de saída<br>
<font color='red'>(b) 4 camadas com 32, 32, 32 e 1 neurônios, 2593 parâmetros e função de ativação relu na camada de saída</font><br>
(c) 5 camadas com 13, 32, 32, 32 e 1 neurônios, 2414 parâmetros e função de ativação sigmóide na camada de saída<br>
(d) 4 camadas com 32, 32, 32 e 1 neurônios, 2593 parâmetros e função de ativação sigmóide na camada de saída<br>

In [9]:
from numpy.random import seed
seed(1)
from tensorflow.random import set_seed
set_seed(2)

from tensorflow.keras.datasets import boston_housing
(x_train, y_train), (x_target, y_target) = boston_housing.load_data()

65536/57026 [==================================] - 0s 0us/step


In [12]:
### seu codigo
model2 = keras.Sequential(
    [
        keras.layers.Dense(32, activation="relu", input_shape=(13,)),
        keras.layers.Dense(32, activation="relu"),
        keras.layers.Dense(32, activation="relu"),
        keras.layers.Dense(1, activation="relu"),
    ]
)
model2.summary()


Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_12 (Dense)             (None, 32)                448       
_________________________________________________________________
dense_13 (Dense)             (None, 32)                1056      
_________________________________________________________________
dense_14 (Dense)             (None, 32)                1056      
_________________________________________________________________
dense_15 (Dense)             (None, 1)                 33        
Total params: 2,593
Trainable params: 2,593
Non-trainable params: 0
_________________________________________________________________


A função 'relu' é a única que não comprime os valores de saída (entre 0 e 1, entre -1 e 1 etc) e deve ser usada.

---

### Exercício 10)

Utilizando a base de dados e o modelo de rede neural criado na questão anterior, compile o modelo utilizando uma função de custo de média quadrática (mse), o otimizador SGD e a taxa de aprendizado 0.01. Adicione a métrica `mae` (mean absolute error)

Depois, normalize os dados (x) por meio da normalização z-score (calcule média e desvio no conjunto de treinamento apenas).

Utilize os dados normalizados para treinar a rede neural por 25 épocas, com batch-size 8. 

Avalie o modelo treinado nos dados de teste, e reporte as posições 0 e 1 do score resultante, respectivamente relativas ao MSE e MAE calculados. Escolha a opção para a qual o intervalo se enquadre nos valores computados.

(a) MSE = (600,630), MAE = (20,25)<br>
(b) MSE = (90,110), MAE = (1,8) <br>
(c) MSE = (3,9), MAE = (50,100) <br>
(d) MSE = (550,570), MAE = (22,24) <br>

In [14]:
# normalizacao z-score
# calculamos média e desvio no treinamento e aplicamos em treinamento e target
mean = x_train.mean(axis=0)
x_train -= mean
std = x_train.std(axis=0)
x_train /= std

x_target -= mean
x_target /= std


# compilamos o modelo
model2.compile(
    optimizer=keras.optimizers.SGD(0.01), loss="mse", metrics=["mae"]
)

batch_size=8
epochs=25
history2 = model2.fit(
    x_train,
    y_train,
    batch_size=batch_size,
    epochs=epochs,
    verbose=1,
    validation_data=(x_target, y_target),
)

Epoch 1/25
51/51 [==============================] - 0s 4ms/step - loss: 347.9305 - mae: 15.6407 - val_loss: 198.9055 - val_mae: 11.5677
Epoch 2/25
51/51 [==============================] - 0s 1ms/step - loss: 128.6384 - mae: 8.2107 - val_loss: 102.1584 - val_mae: 7.3856
Epoch 3/25
51/51 [==============================] - 0s 1ms/step - loss: 90.7606 - mae: 6.6279 - val_loss: 87.1860 - val_mae: 6.5934
Epoch 4/25
51/51 [==============================] - 0s 1ms/step - loss: 85.5814 - mae: 6.5865 - val_loss: 85.1194 - val_mae: 6.5212
Epoch 5/25
51/51 [==============================] - 0s 1ms/step - loss: 84.9476 - mae: 6.5702 - val_loss: 83.9496 - val_mae: 6.5238
Epoch 6/25
51/51 [==============================] - 0s 994us/step - loss: 84.7818 - mae: 6.6279 - val_loss: 83.7019 - val_mae: 6.5334
Epoch 7/25
51/51 [==============================] - 0s 2ms/step - loss: 84.7615 - mae: 6.6271 - val_loss: 83.5028 - val_mae: 6.5466
Epoch 8/25
51/51 [==============================] - 0s 2ms/step - lo

In [16]:
score = model2.evaluate(x_target, y_target, verbose=0)
score

[83.7920150756836, 6.519824981689453]

In [17]:
score = model2.evaluate(x_target, y_target, verbose=0)
print("MSE: %.1f" % (score[0]))
print("MAE: %.1f" % (score[1]))

MSE: 83.8
MAE: 6.5
